---
title: Data Preparation and Quality Assessment in Genome Assembly
type: workshop
display: basic
no-caldate: true
provider: ISU
hideprovider: true
description: This workshop provides a hands-on introduction to data preparation, genome assembly, and quality assessment.
categories: [2025 Bioinfo]
parent: 
  title: Bioinformatics Workshop Series
  url: /events/2025-bioinfo


layout_type: workshop
has-sessions: true
time: 1 â€“ 5 PM ET
registration:
    url: https://forms.office.com/g/8qFLk99g5c
tags: bioinformatics

prerequisites:
  - text: Familiarity with basic command-line concepts. 
---

This workshop provides a hands-on introduction to data preparation, genome assembly, and quality assessment. Participants will explore different assembly approaches and techniques for evaluating the accuracy and completeness of genome assemblies, helping attendees understand key metrics and statistical methods used to assess the quality of genomic data.  <!--excerpt-->


## Data Preparation
by Viswanathan Satheesh and Sivanandan Chudalayandi 


**Load Modules** 

```bash 

module load merqury 

module load fastqc 

module use /software/el9/spack/lmod/Core 

module load jellyfish 

``` 

 

**Activate virtual environment for NanoPlot** 

```bash 

source /project/scinet_workshop2/Bioinformatics_series/virtual_env/bin/activate 

``` 

Data directories: 01_Data and 02_References 

 

Data QC - Illumina and PacBio HiFi reads 

 

### Illumina - `FastQC` 

 

```bash 

mkdir 03_IlluminaQC 

fastqc -o 03_IlluminaQC -t 2 01_Data/AT_Illumina_paired_*fastq 

``` 

Estimated Time: 1m 8s 

 

 

#### Per base sequence quality 

![](assets/images/fastqc/per_base_quality.png) 

 

#### Per base sequence content 

![](assets/images/fastqc/per_base_sequence_content.png) 

 

#### Overrepresented sequences 

![](assets/images/fastqc/duplication_levels.png) 

 

#### Adapter content 

![](assets/images/fastqc/adapter_content.png) 

 

### PacBio HiFi - `nanoplot` 

```bash 

mkdir 04_NanoPlotQC 

time NanoPlot --fastq 00_Raw_Data/mapped_reads.chr2.filtlong.fastq.gz -o 04_NanoPlotQC --threads 8 

``` 

Estimated Time: 2m 17s 

<pre> 

real    1m31.985s 

</pre> 

 

It has taken `NanoPlot` about 2  minute and 17 seconds to run. 

 

[View Report](assets/images/03_QC_Nanoplot/NanoStats.txt) 

 

![](assets/images/03_QC_Nanoplot/LengthvsQualityScatterPlot_dot.png) 

 

![](assets/images/03_QC_Nanoplot/LengthvsQualityScatterPlot_kde.png) 

 

![](assets/images/03_QC_Nanoplot/Non_weightedHistogramReadlength.png) 

 

![](assets/images/03_QC_Nanoplot/Non_weightedLogTransformed_HistogramReadlength.png) 

 

![](assets/images/03_QC_Nanoplot/WeightedHistogramReadlength.png) 

 

![](assets/images/03_QC_Nanoplot/WeightedLogTransformed_HistogramReadlength.png) 

 

![](assets/images/03_QC_Nanoplot/Yield_By_Length.png) 

 

### Predicting the genome size with Illumina reads with GenomeScope 

 

```bash 

mkdir 05_GenomeScope 

time jellyfish count -m 21 -s 100M -t8 \ 

-C 01_Data/AT_Illumina_paired_* \ 

-o 05_GenomeScope/reads.jf 

``` 

 

- `-m 21`: 21-mers 

- `-s 10M`: 10 million reads # this is the number of reads used to estimate the size of the genome 

- `-t 8`: 8 threads 

 

Takes about 1m 50 seconds to run. 

 

```bash 

time jellyfish histo \ 

  -t 8 05_GenomeScope/reads.jf > 05_GenomeScope/reads.histo 

``` 

 

- `jellyfish histo` is used to estimate the size of the genome using the `reads.jf` file. It outputs the size of the genome in the `reads.histo` file. 

 

Takes about 20 seconds to run. 

 

We can now copy the `reads.histo` to our local computer and upload it to [GenomeScope](http://genomescope.org/) 

 

### Interpretation of the images: 

 

![](assets/images/genome_scope.png) 

 

_Common Elements in Both Figures_ 

 

- Genome Size (len): Estimated genome size is 21,873,679 bp (12.53 Mb). 

- Unique Sequence (uniq): 82.8% of the genome is unique sequence. 

- Heterozygosity (het): The heterozygosity rate is 0.0829%, indicating a very low level of heterozygosity. 

- Coverage (kcov): Average k-mer coverage is 29.6. 

- Error Rate (err): Estimated sequencing error rate is 0.38%. 

- Duplication (dup): 1.51% of the genome is duplicated. 

- k-mer size (k): k-mer length used for the analysis is 21. 

 

_Key Elements in the Plots_ 

 

- X-Axis (Coverage): Represents the k-mer coverage. In the linear plot, it is shown on a linear scale, whereas in the logarithmic plot, it is shown on a logarithmic scale. 

- Y-Axis (Frequency): Represents the frequency of k-mers at different coverage levels. 

- Blue Bars (observed): Histogram of observed k-mer frequencies. 

- Black Line (full model): Model fit to the observed k-mer frequencies. 

- Yellow Line (unique sequence): Contribution of unique sequences to the k-mer frequencies. 

- Orange Line (errors): Contribution of sequencing errors to the k-mer frequencies. 

- Dashed Lines (kmer-peaks): Peaks corresponding to k-mer coverage of unique and repetitive sequences. 

- Red Dashed Line (cov-threshold): A threshold to distinguish high-coverage k-mers, typically used to identify potential contaminant sequences or highly repetitive regions. This is set at a very high coverage level (around 1000). 
