---
title: Introduction to Bioinformatics
type: workshop
display: basic
no-caldate: true
provider: ISU
instructor: "Lavida Rogers (SCINet Office, USDA-ARS), Siva Chudalayandi, Rick Masonbrink, and Viswanathan Satheesh (Genome Informatics Facility, ISU)"
date: 2025-04-15
hideprovider: true
description: This workshop provides an overview of bioinformatics covering applications, sequencing technologies, and basic workflows.
categories: [2025 Bioinfo]
parent: 
  title: Bioinformatics Workshop Series
  url: /events/2025-bioinfo

layout_type: workshop
has-sessions: true
time: 1 – 4 PM ET
registration:
    url: https://forms.office.com/g/8qFLk99g5c

tags: bioinformatics

prerequisites:
  - text: Familiarity with basic command-line concepts. 
---

This workshop provides an overview of bioinformatics covering applications, sequencing technologies, and basic workflows. Participants will also explore various file formats in bioinformatics and gain hands-on experience using NCBI’s BLAST web tool to perform sequence alignments and interpret results.  <!--excerpt-->

## Tutorial Setup Instructions 

Steps to prepare for the tutorial session each day: 

* **Logging on to [Ceres Open OnDemand (OOD)](http://ceres-ood.scinet.usda.gov/):** Please confirm you can successfully log in to Atlas OOD with your SCINet account [(see instructions here)]({{site.baseurl}}/guides/access/web-based-login). If you are successful, you will be able to see the Ceres OOD home page.
* **Ceres Shell Access:** When on Ceres OOD, click on the top navigation bar: “Clusters” > “Ceres Shell Access”. A new tab will appear that looks like a shell terminal (e.g., like PowerShell). Please confirm you do not receive any error messages or requests to re-authenticate and that the final line looks like "[firstname.lastname@ceres ~]$” 
* **RStudio Server:** Back on the main Ceres OOD tab, click on the top or side navigation bar: “Interactive Apps” > “RStudio Server”. 
  * Fill the input fields with the following (input fields not listed below can be left at their default values):  
    * Account: scinet_workshop2 
    * Queue: ceres 
    * QOS: 400thread 
    * R Version: 4.4.1 
    * Number of hours: 1 
    * Number of cores: 1 
    * Memory required: 8GB 
    * Optional Slurm Arguments: (leave empty) 
  * Click the "Launch" button. 
  * Wait a moment for the job card to update from "Queued" to "Running". 
  * Please confirm that clicking on the “Connect to RStudio Server” button opens a new tab with the RStudio Server interface. 

-----  

## Tutorial Introduction

In this tutorial we will use the command line and other bioinformatics tools to explore the different types of data you will encounter in bioinformatics. The exercises and examples provided will guide you through understanding file structures, examining sequences and their associated quality scores and parsing/filtering variant data. We will also explore public databases and use BLAST for performing sequence alignment.



**Throughout the tutorial we will use the following command-line tools:**
- Basic text processing: `grep`, `cut`, `bioawk`, `wc`
- Quality control: `fastqc`
- Filtering variants: `bcftools`
- Downloading public data: `wget`
- Raw sequences from NCBI SRA: `SRA Toolkit`

-----
## Exploring Bioinformatics File Formats

{% include table caption="Common file formats in bioinformatics" content="|Format | Description | Use case|File extensions|
|--------|------------|---------|---------------|
|FASTA | text file with nucleotide or protein sequences | stores raw sequence data| .fa, .fasta,.fna|
| FASTQ | text file with sequences and quality scores | next-generation sequencing reads| .fq, .fastq|
|SAM/BAM| sequence alignment map (SAM) and binary alignment map (BAM)| stores aligned reads | .sam, .bam|
|VCF| variant call format | stores genomic variants (e.g. SNPs)| .vcf|
|GFF/GTF| 9 column text format with genomic features| stores information about gene annotation|.gff, .gtf|" %}

### Part 1: Exploring FASTA Files
FASTA files contain a single-line description and ID followed by one or more lines of sequence data. 

*Line 1: starts with “>” followed by ID*  
*Line 2: Sequence data*

**Let's take a look at an actual file:**
1. Are you in the right working directory? 

    {:.copy-code}
    ```bash
cd /90daydata/shared/$USER/intro_bioinformatics /day1 
```

1. We can look at the beginning of the file by using `head`, which displays the first few lines of files. 

    {:.copy-code}
    ```bash
head files/GCF_000001735.4_TAIR10.1_genomic.fna 
```  

1. We can also view the first 4 lines using `-n`: 

    {:.copy-code}
    ```bash
head -n 4 files/GCF_000001735.4_TAIR10.1_genomic.fna
```  

1. Another way to view a fasta file is to use `cat` which is used to read and display file contents. 

    {:.copy-code}
    ```bash
cat files/GCF_000001735.4_TAIR10.1_genomic.fna #this will display the entire file and should only be used for small files 
``` 

    {:.copy-code}
    ```bash
cat -n 4 files/GCF_000001735.4_TAIR10.1_genomic.fna
```  

1. We can also look at the last few lines of the file by using `tail`: 

    {:.copy-code}
    ```bash
tail files/GCF_000001735.4_TAIR10.1_genomic.fna
```  

1. For a scrollable preview of a file we can use less  

    {:.copy-code}
    ```bash
less files/ GCF_000001735.4_TAIR10.1_genomic.fna
```

{% include alert no_icon=true type="success" title="You try:" text="How would you look at the last 20 lines of the file?" %}


#### Counting the number of sequences with `grep`:  
`grep` is a search tool that is commonly used to find lines that match a pattern.  

{:.copy-code}
```bash
grep -c "^>" files/GCF_000001735.4_TAIR10.1_genomic.fna
```

- `grep`: used for matching lines
- `-c`: count; prints number of matching lines
- `"^>"`: regular expression pattern that matches any line that starts with `>`

#### Determine sequence lengths with `bioawk`:  

<div class="bg-primary-lighter padding-3 margin-3 radius-md" markdown="1"> 

{:.margin-0.font-heading-md}
##### Quick Guide on `bioawk`
-----
`bioawk` is an extension of `awk`, which is a tool used for parsing and processing text files. `bioawk` has specialized features to handle biological file formats.

{% include table fixed=true content="|`awk`|`bioawk`|
|-----|--------|
|general text processing| specialized for biological files|
|does not understand the format of bioloical files| aware of biological file formats|
|manually split fields that are identified only by number (e.g., `$1`, `$2`)| Automatically generates named fields from file contents (e.g., `$seq`, `$qual`)|
|requires extra code/logic to parse biological data| has built-in functions for parsing sequences, reads and variants|" %}

{% include table content="|Command| Description|Fields|
|-------|------------|--------|
|`bioawk -c fasta`| parses sequences (FASTA)|`$name`, `$seq`|
|`bioawk -c fastx`| parses sequencing reads with quality scores (FASTQ)|`$name`, `$seq`, `$qual`|
|`bioawk -c vcf`| parses genetic variants (VCF)|`$chrom`, `$pos`,`$ref`,`$qual`, etc.|" %}

**Basic format of a `bioawk` command:
```
bioawk -c <file format> ‘action’ input_file
```
</div>

**Get sequence names and lengths**

{:.copy-code}
```bash
module load bioawk
bioawk -c fasta '{print $name length($seq)}' files/GCF_000001735.4_TAIR10.1_genomic.fna | head
```

***Note:*** *The pipe `|` in the line of code above is taking the output of the command on the left and using it as the input for the command on the right.*

***Note:*** *The `{}` is telling bioawk to perform the action provided on each line/record*

{% include alert no_icon=true type="success" title="You try:" text="Print sequences with lengths longer than 100 bp" %}

{% include alert no_icon=true type="info" title="Discussion Questions:" text="<ul><li>Why is it important to look at sequence lengths?</li>
<li>What could the sequence header information be used for in analysis downstream?</li></ul>" %}


### Part 2: Exploring FastQ files
FastQ files are similar to fasta files, but also contain the quality score of the sequence data. 

***Line 1: starts with “@” followed by ID  
Line 2: Sequence data  
Line 3: Starts with “+”    
Line 4:Quality score for each base in the sequence***

Quality scores indicate the confidence of the base call by the sequencer. 

- Quality scores are encoded as American Standard Code for Information Interchange (ASCII) characters. 
- Different types of encodings are available and vary across sequencing technologies 

#### Phred quality score \(Q\) calculation: 
\[  
    Q= -10 \times \log_{10}(P)  
\]  

Where:   
- \(Q\) = Phred quality score
- \(P\) = probability that base call is incorrect

{% include table fixed=true content="|ASCII Char| Phred Score | Error Probability|
|----------|-------------|-------------------|
|! | 0 | 1.00|
|I| 40| 0.0001|
|@| 31 | 0.00079|" %}

{% include table fixed=true content="
|Phred Score| Error Probability | Interpretation|
|-----------|-------------------|---------------|
|10| 1 in 10 | 90% accurate|
|20|1 in 100| 99% accurate|
|30| 1 in 1000| 99.9% accurate|
|40| 1 in 10000| 99.99% accurate|" %}

**Let's take a look at an actual file:**

{:.copy-code}
```bash
head files/SRR4420293_1.fastq
```

```bash
head -n 4 files/SRR4420293_1.fastq
```

We can also use `bioawk`:  

As a reminder of the format of bioawk commands: `bioawk -c <file format> ‘action’ input_file`

```bash
bioawk -c fastx '{print $name, length($seq), $qual}' files/SRR4420293_1.fastq | head -n 5
```

***Note:*** *The pipe `|` in the line of code above is taking the output of the command on the left and using it as the input for the command on the right.*



{% include alert no_icon=true type="success" title="You try:" text="Count reads in the FASTQ file using <code>grep</code>" %}


#### Other ways to count the number of reads: 

* **Count lines using wc:**

  {:.copy-code}
  ```bash
wc -l files/SRR4420293_1.fastq
``` 

  ***But this gives line counts. Is that what we want? What should we do here to get the number of reads?***

  1. First let’s get the number of lines without the file name: 

      {:.copy-code}
      ```bash
wc -l < files/SRR4420293_1.fastq
```  
      “<” here is feeding the contents of the file into wc without passing the filename as an argument. This is called input redirection.  
      <br>
  2. In order for us to do integer math directly in bash we need: `$((….))` and for us to see the result of the calculation in the shell we will use echo. 

      {:.copy-code}
      ```bash
echo $(( $(wc -l < files/SRR4420293_1.fastq)/4)) 
```

  3. We also need to put the wc function in `$()`, which will cause bash to execute the command inside the parentheses and then use the output when evaluating the rest of the command: 

      {:.copy-code}
      ```bash
echo $(( $(wc -l < files/SRR4420293_1.fastq)/4)) 
```

      Let us break down the code: 
        - `$(wc -l < files/SRR4420293_1.fastq)`: to get line count
        -  `$((....))`: Do integer math directly in Bash:
        - `echo`: prints the result   
        <br>


* **Using bioawk:**   

    {:.copy-code}
    ```bash
bioawk -c fastx '{print $name}' files/SRR4420293_1.fastq | wc -l
```

    {:.copy-code}
    ```bash
bioawk -c fastx 'END {print NR}' files/SRR4420293_1.fastq
```

    **What's the average read length?** 

    {:.copy-code}
    ```bash
bioawk -c fastx '{sum += length($seq)} END {print sum/NR} files/SRR4420293_1.fastq
```

      - `sum += length($seq)` keeps adding each read's length to a running total 
      - `NR` - a built in variable in `bioawk` that stands for the number of records/reads
      - `END`: end of the file

### Let's look at the file quality using FastQC
FastQC is a tool used to check the quality of sequencing reads from FASTQ files. The report tells us about base quality, sequence length distributions, adapter content, GC content and overrepresented sequences.

{:.copy-code}
```bash
module load fastqc
fastqc -o fastqc_output files/SRR4420293_1.fastq
```

`-o`: output directory

{% include alert no_icon=true type="info" title="Discussion Questions:" text="<ul><li>What trends do you observe in quality across reads?</li>
<li>Which reads should we consider trimming or filtering?</li>
<li>How could poor-quality reads impact downstream analysis?</li></ul>" %}

### Part 3: Exploring Variant Call Format (VCF) Files
These files store sequence variants, SNPs and indels. 

- Metadata lines: each line starts with ## followed by key=value pairs
- single header line : starts with single # and describes columns in the data lines
- data lines: sequence variation data

#### Layout and structure of VCF files:

* **View the header only** 

    {:.copy-code}
    ```bash
grep '^#' files/chinook.vcf | less
```

* **View the first few variant lines:** 

    {:.copy-code}
    ```bash
grep -v '^#' files/chinook.vcg | less
```
    Note: `-v` shows lines that do not match the pattern provided

* **Inspecting columns in VCF files**
    The `cut` command is used to extract specific columns/fields from a file.  

* **What's in the INFO field?** 

    {:.copy-code}
    ```bash
cut -f8 files/chinook.vcf | grep -v '^#' | head -n 5
```

* **Let's look at the kinds of mutations** 

    {:.copy-code}
    ```bash
cut -f5 files/chinook.vcf | grep -v '^#' | sort | uniq -c | sort -nr| head ```

* **What chromosomes are in this file?** 

    {:.copy-code}
    ```bash
cut -f1 chinook.vcf | grep -v '^#' | sort | uniq
```


### Look at a VCF file using `bcftools` and `bioawk`:

`bcftools` is a command-line tool used for viewing, filtering and manipulating VCF files. We will use `bcftools` to help us summarize variant info and for extracting and selecting variants. 

{% include table caption="Options associated with bcftools" content="|Option|Explanation|
|-------|------------|
|-r|the output is restricted to a specific region|
|i| include records that satisfy the given condition|
|-v| filter to show the variant type |
|-H| skips the VCF header lines and prints the variant lines only|" %}

* **Using `bcftools`:** 

    {:.copy-code}
    ```bash
module load bcftools
    `bcftools view files/chinook.vcf | head -n 20
```

* **Using `bioawk`:** 

    {:.copy-code}
    ```bash
bioawk -c vcf '{print $chrom, $pos, $ref, $alt, $qual}' files/chinook.vcf | head -n 5
```

* **Count the total number of variants:** 

    {:.copy-code}
    ```bash
bcftools view -H files/chinook.vcf| wc -l
```

    {:.copy-code}
    ```bash
bioawk -c vcf 'END {print NR}' files/chinook.vcf
```

* **Filtering variants**
  * By quality:  

      {:.copy-code}
      ```bash
bcftools view -i '%QUAL>=30' files/chinook.vcf > high_qual_bcf.vcf
``` 
      {:.copy-code}
      ```bash
bcftools view -i 'QUAL>100' files/chinook.vcf > high_qual_bcf.vcf 
```
      
      {:.copy-code}
      ```bash
head high_qual_bcf.vcf
``` 

      {:.copy-code}
       ```bash
bioawk -c vcf '$qual >=30' files/chinook.vcf > high_qual_bioawk.vcf
``` 

      {:.copy-code}
      ```bash
head high_qual_bioawk.vcf
```

  * By variant type: 

      {:.copy-code}  
      ```bash
bioawk -c vcf 'length($ref) == 1 && length($alt) ==1' > snps_only_bioawk.vcf
``` 

      {:.copy-code}
      ```bash
bcftools view -v snps files/chinook.vcf -0z -o snps_only_bcf.vcf
```

  * By chromosome: 

      {:.copy-code}
      ```bash
bcftools view -r chr5 files/chinook.vcf > chr5_variants.vcf
``` 

      {:.copy-code}
      ```bash
bioawk -c vcf '$chrom == "chr5"' files/chinook.vcf > chr5_only_bioawk.vcf
```

* **VCF file summary with `bcftools`**  
  `bcftools` is able to generate a summary statistics file on your VCF files: 

    {:.copy-code}
    ```bash
bcftools stats files/chinook.vcf > stats_vcf.txt
``` 

    {:.copy-code}
    ```bash
less stats_vcf.txt
```

  * To plot: 

      {:.copy-code}
      ```bash
plot-vcfstats -p stats_output stats_vcf.txt
```


{% include alert no_icon=true type="success" title="You try:" text="How many SNPs pass the quality threshold?" %}

{% include alert no_icon=true type="info" title="Discussion Questions:" text="<ul><li>Why is filtering by quality important?</li>
<li>Why might SNPs be prioriticed over other variant types?</li>
<li>How would you decide which variants are biologically relevant?</li>
<li>How might filters change depending on if you were studing diesease vs breeding?</li></ul>" %}


## Exploring public repositories/databases 

{:.copy-code}
```bash
wget FTP link
```

**Download data with SRA toolkit**
 
{:.copy-code}
```bash
module load sra-tools fasterq-dump
```

{% include alert no_icon=true type="success" title="You try:" text="<ul><li>Choose a database and search for a dataset of interest.</li>
<li>Download one FASTQ, FASTA or VCF file and use some of the commands discussed to examine the file.</li>
<li>Share something you learned about your dataset. For example: How many reads or variants are present in the file?</li></ul>"%}

## BLAST

## Sources

- Bioinformatics Workbook [Online] Available at: https://bioinformaticsworkbook.org/#gsc.tab=0 (Accessed April 5, 2025)
- Anderson, E.C (2024). Practical Computing and Bioinformatics for Conservation and Evolutionary Genomics. Accessed April 11, 2025

